packages/backend-common/src/service/lib/hostFactory.ts

```diff
@@ -45,48 +50,152 @@ export function createHttpServer(
  * @returns A Https server instance
  *
  */
-export function createHttpsServer(
+export async function createHttpsServer(
   app: express.Express,
   httpsSettings: HttpsSettings,
   logger?: Logger,
-): http.Server {
+): Promise<http.Server> {
   logger?.info('Initializing https server');
 
-  const credentials: { key: string; cert: string } = {
-    key: '',
-    cert: '',
-  };
+  let credentials: { key: string | Buffer; cert: string | Buffer };
 
   const signingOptions: any = httpsSettings?.certificate;
 
-  if (signingOptions?.algorithm !== undefined) {
-    logger?.info('Generating self-signed certificate with attributes');
+  // TODO(Rugvip): remove support for generated certificate params and make this a more straightforward check
+  if (signingOptions?.attributes) {
+    credentials = await getGeneratedCertificate(signingOptions, logger);
+  } else {
+    logger?.info('Loading certificate from config');
 
-    const certificateAttributes: Array<any> = Object.entries(
-      signingOptions.attributes,
-    ).map(([name, value]) => ({ name, value }));
+    credentials = {
+      key: signingOptions?.key,
+      cert: signingOptions?.cert,
+    };
+  }
 
-    // TODO: Create a type def for selfsigned.
-    const signatures = require('selfsigned').generate(certificateAttributes, {
-      algorithm: signingOptions?.algorithm,
-      keySize: signingOptions?.size || 2048,
-      days: signingOptions?.days || 30,
-    });
+  if (!credentials.key || !credentials.cert) {
+    throw new Error('Invalid HTTPS credentials');
+  }
 
-    logger?.info('Bootstrapping self-signed certificate');
+  return https.createServer(credentials, app) as http.Server;
+}
 
-    credentials.key = signatures.private;
-    credentials.cert = signatures.cert;
+async function getGeneratedCertificate(
+  options: CertificateSigningOptions,
+  logger?: Logger,
+) {
+  if (options?.algorithm) {
+    logger?.warn(
+      'Certificate generation configuration with parameters in backend.https.certificate is deprecated, set backend.https = true instead',
+    );
+  }
+
+  const hasModules = await fs.pathExists('node_modules');
+  let certPath;
+  if (hasModules) {
+    certPath = resolvePath(
+      'node_modules/.cache/backstage-backend/dev-cert.pem',
+    );
+    await fs.ensureDir(dirname(certPath));
   } else {
-    logger?.info('Bootstrapping cert from config');
+    certPath = resolvePath('.dev-cert.pem');
+  }
 
-    credentials.key = signingOptions?.key;
-    credentials.cert = signingOptions?.cert;
+  let cert = undefined;
+  if (await fs.pathExists(certPath)) {
+    const stat = await fs.stat(certPath);
+    const ageMs = Date.now() - stat.ctimeMs;
+    if (stat.isFile() && ageMs < ALMOST_MONTH_IN_MS) {
```

Comment:
should you do `daysToMs((options?.days || 30) - 5)` instead?

Reply:
Nah, old config is deprecated so it's hardcoded to 30 days